### PR TITLE
8306706: Support out-of-line code generation for MachNodes

### PR DESCRIPTION
**Pull request by merykitty**
Hi,

This patch adds supports for MachNodes to emit an out-of-line piece of code in the stub section of the compiled method. This allows the separation of the uncommon path from the common one, which speeds up the common path a little bit and increases compiled code density. Please take a look and leave reviews.

Thanks a lot.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8306706](https://bugs.openjdk.org/browse/JDK-8306706): Support out-of-line code generation for MachNodes

### Reviewers
 * [Tobias Hartmann](https://openjdk.org/census#thartmann) (@TobiHartmann - **Reviewer**) ⚠️ Review applies to [a17bcb76](https://git.openjdk.org/jdk/pull/13602/files/a17bcb76bac7bb207728fbe73a80eae3087b5c1c)
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/13602/head:pull/13602` \
`$ git checkout pull/13602`

Update a local copy of the PR: \
`$ git checkout pull/13602` \
`$ git pull https://git.openjdk.org/jdk.git pull/13602/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 13602`

View PR using the GUI difftool: \
`$ git pr show -t 13602`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/13602.diff">https://git.openjdk.org/jdk/pull/13602.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/13602#issuecomment-1519127252)

### CODE REVIEW

**Comment by merykitty** [comment]
With this patch, the compiled code for a float-to-int conversion is changed:

    Before:

        vcvttss2si %xmm1,%eax
        cmp    $0x80000000,%eax
        jne    DONE
        sub    $0x8,%rsp
        vmovss %xmm1,(%rsp)
        call   Stub::f2i_fixup              ;   {runtime_call StubRoutines (initial stubs)}
        pop    %rax
    DONE:


    After:

        vcvttss2si %xmm1,%eax
        cmp    $0x80000000,%eax
        je     STUB
    CONTINUE:

    STUB:
        sub    $0x8,%rsp
        vmovss %xmm1,(%rsp)
        call   Stub::f2i_fixup              ;   {runtime_call StubRoutines (initial stubs)}
        pop    %rax
        jmp    CONTINUE

And there are slight improvements shown in microbenchmarks, although the result differs run-to-run, the patched version seems to be generally more performant:

                                          Before             After
    Benchmark             Mode  Cnt    Score    Error    Score    Error  Units   Change
    ConvertF2I.d2iArray   avgt    5  266.890 ±  3.277  260.720 ±  1.382  ns/op   -2.31%
    ConvertF2I.d2iSingle  avgt    5    0.378 ±  0.005    0.317 ±  0.013  ns/op  -16.14%
    ConvertF2I.d2lArray   avgt    5  273.999 ± 12.571  267.862 ±  4.806  ns/op   -2.24%
    ConvertF2I.d2lSingle  avgt    5    0.379 ±  0.005    0.348 ±  0.044  ns/op   -8.18%
    ConvertF2I.f2iArray   avgt    5  261.549 ±  1.391  255.522 ± 15.133  ns/op   -2.30%
    ConvertF2I.f2iSingle  avgt    5    0.378 ±  0.005    0.311 ±  0.007  ns/op  -17.72%
    ConvertF2I.f2lArray   avgt    5  272.745 ±  1.661  267.770 ±  7.033  ns/op   -1.82%
    ConvertF2I.f2lSingle  avgt    5    0.379 ±  0.007    0.350 ±  0.022  ns/op   -7.65%

**Comment by merykitty** [comment]
The generated node for the stub looks like this:

    class convF2I_reg_regStub : public C2CodeStub {
    private:
        const convF2I_reg_regNode* _node;
        PhaseRegAlloc* ra_;
        MachOper* opnd_array(uint index) const { return _node->opnd_array(index); }
    public:
        convF2I_reg_regStub(const convF2I_reg_regNode* node, PhaseRegAlloc* ra) : _node(node), ra_(ra) {}
        int max_size() const { return 23; }
        void emit(C2_MacroAssembler& masm);
    };

And the corresponding node's `emit` method has an additional section:

    void convF2I_reg_regNode::emit(CodeBuffer& cbuf, PhaseRegAlloc* ra_) const {
        cbuf.set_insts_mark();
        convF2I_reg_regStub* stub = new (Compile::current()->comp_arena()) convF2I_reg_regStub(this, ra_);
        if (!Compile::current()->output()->in_scratch_emit_size()) {
            Compile::current()->output()->add_stub(stub);
        }
        ...
    }

**Review by TobiHartmann — COMMENTED** [review]
Great work. I'm just wondering if the extra complexity is justified for optimizing only the floating point conversions. Do you plan to use this for other optimizations?

**Comment by merykitty** [comment]
@TobiHartmann Thanks for taking a look, I think this can be used for the vectorized version of these nodes, as well as the max, min nodes for floating point numbers. I also see compact header uses out-of-line code to slow path `LoadNKlass`.

**Review by vnkozlov — COMMENTED** [review]
Is it possible to do this in `c2_MacroAssembler_x86` instead (as for `verified_entry`)?
We are trying to move complex coding from .ad files to macroassembler.

**Comment by merykitty** [comment]
@vnkozlov Yes we can explicitly define a stub without relying on code generation, it may be more preferable since it avoids adding complexity to adlc generation. The only downside is that there is some boilerplate for each usage but I think the boilerplate is not too terrible.

**Comment by vnkozlov** [comment]
> @vnkozlov Yes we can explicitly define a stub without relying on code generation, it may be more preferable since it avoids adding complexity to adlc generation. The only downside is that there is some boilerplate for each usage but I think the boilerplate is not too terrible.

Can you look on that? There could be other cases in Macroassembler which can use this

**Review by vnkozlov — COMMENTED** [review]
Clever. Let me test it.

**Review by vnkozlov — APPROVED** [review]
My testing passed. Good.

File: src/hotspot/share/opto/c2_CodeStubs.hpp [hunk 1]
```diff
@@ -101,4 +112,32 @@ class C2HandleAnonOMOwnerStub : public C2CodeStub {
 };
 #endif
 
+template <class... Ts>
+class C2GeneralStub : public C2CodeStub {
```
Comment (TobiHartmann) [inline src/hotspot/share/opto/c2_CodeStubs.hpp:116]: A comment describing the purpose of this class would be nice.

**Review by TobiHartmann — APPROVED** [review]
Nice! The new version looks good to me.

**Comment by merykitty** [comment]
@vnkozlov Thanks for your reviews and testing
@TobiHartmann Thanks for your suggestion, I have added comments to describe the purpose of `C2GeneralStub`

**Review by vnkozlov — APPROVED** [review]
Good.

File: src/hotspot/cpu/x86/c2_MacroAssembler_x86.cpp [hunk 2]
```diff
@@ -4293,6 +4293,56 @@ void C2_MacroAssembler::arrays_equals(bool is_array_equ, Register ary1, Register
...
+      cmp64(dst, ExternalAddress(StubRoutines::x86::double_sign_flip()));
+      slowpath_target = StubRoutines::x86::f2l_fixup();
+    } else {
+      cvttsd2siq(dst, src);
+      cmp64(dst, ExternalAddress(StubRoutines::x86::double_sign_flip()));
+      slowpath_target = StubRoutines::x86::d2l_fixup();
+    }
+  }
+
+  auto stub = C2CodeStub::make<Register, XMMRegister, address>(dst, src, slowpath_target, 23, convertF2I_slowpath);
```
Comment (vamsi-parasa) [inline src/hotspot/cpu/x86/c2_MacroAssembler_x86.cpp:4339]: Hi @merykitty, could you please explain how the size 23 was computed? This value does not work with APX and I created a PR (https://github.com/openjdk/jdk/pull/25787) for that.

Comment (merykitty) [reply src/hotspot/cpu/x86/c2_MacroAssembler_x86.cpp:4339]: @vamsi-parasa Hi, I just manually assembled the snippet and see its size, for such a small snippet it is easy to see that the size is indeed the largest possible.

**Pull request closed without merging**